File: parser/src/main/java/com/github/fburato/highwheelmodules/bytecodeparser/classpath/CompoundClassPathRoot.java
package com.github.fburato.highwheelmodules.bytecodeparser.classpath;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import com.github.fburato.highwheelmodules.model.classpath.ClasspathRoot;
import com.github.fburato.highwheelmodules.model.bytecode.ElementName;

/**
 * Wraps multiple child roots.
 */
public class CompoundClassPathRoot implements ClasspathRoot {

    private final List<ClasspathRoot> roots;

    public CompoundClassPathRoot(final List<ClasspathRoot> roots) {
        this.roots = roots;
    }

    public InputStream getData(final ElementName name) throws IOException {
        for (final ClasspathRoot each : this.roots) {
            final InputStream is = each.getData(name);
            if (is != null) {
                return is;
            }
        }
        return null;
    }

    public Collection<ElementName> classNames() {
        final List<ElementName> cns = new ArrayList<>();
        for (final ClasspathRoot each : this.roots) {
            cns.addAll(each.classNames());
        }
        return cns;
    }

    public InputStream getResource(final String name) throws IOException {
        for (final ClasspathRoot each : this.roots) {
            final InputStream is = each.getResource(name);
            if (is != null) {
                return is;
            }
        }
        return null;
    }

}

File: parser/src/main/scala/com/github/fburato/highwheelmodules/bytecodeparser/classpath/CompoundClassPathRoot.scala
package com.github.fburato.highwheelmodules.bytecodeparser.classpath

import com.github.fburato.highwheelmodules.model.bytecode.ElementName
import com.github.fburato.highwheelmodules.model.classpath.ClasspathRoot
import com.github.fburato.highwheelmodules.utils.TryUtils._

import java.io.InputStream
import scala.util.Try

class CompoundClassPathRoot(roots: Seq[ClasspathRoot]) extends ClasspathRoot {
  override def getData(elementName: ElementName): Try[Option[InputStream]] =
    sequence(roots.map(cpr => cpr.getData(elementName)))
      .map(getFirstNonNullOrNull)

  private def getFirstNonNullOrNull(streams: Seq[Option[InputStream]]): Option[InputStream] =
    streams
      .flatMap(maybeStream => maybeStream.toSeq)
      .headOption

  override def classNames: Try[Seq[ElementName]] =
    sequence(roots.map(cpr => cpr.classNames))
      .map(result => result.flatten)

  override def getResource(name: String): Try[Option[InputStream]] =
    sequence(roots.map(cpr => cpr.getResource(name)))
      .map(getFirstNonNullOrNull)
}
